extras/rya.pcj.fluo/pcj.fluo.app/src/main/java/org/apache/rya/indexing/pcj/fluo/app/query/FluoQueryMetadataCache.java

```diff
@@ -0,0 +1,247 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */package org.apache.rya.indexing.pcj.fluo.app.query;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.concurrent.Callable;
+
+import org.apache.fluo.api.client.SnapshotBase;
+import org.apache.fluo.api.data.Bytes;
+import org.apache.fluo.api.data.Column;
+import org.apache.rya.indexing.pcj.fluo.app.NodeType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Optional;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+
+/**
+ * Wrapper for {@link FluoQueryMetadataDAO} that caches any metadata that has been retrieved from Fluo. This class first
+ * checks the cache to see if the metadata is present before delegating to the underlying DAO method to retrieve the
+ * data.
+ *
+ */
+public class FluoQueryMetadataCache extends FluoQueryMetadataDAO {
+
+    private static final Logger LOG = LoggerFactory.getLogger(FluoQueryMetadataCache.class);
+    private final FluoQueryMetadataDAO dao;
+    private final Cache<String, CommonNodeMetadata> commonNodeMetadataCache;
+    private final Cache<String, Bytes> metadataCache;
+    private int capacity;
+    private int concurrencyLevel;
+
+    /**
+     * Creates a FluoQueryMetadataCache with the specified capacity. Old, unused results are evicted as necessary.
+     *
+     * @param capacity - max size of the cache
+     */
+    public FluoQueryMetadataCache(FluoQueryMetadataDAO dao, int capacity, int concurrencyLevel) {
+        this.dao = dao;
+        commonNodeMetadataCache = CacheBuilder.newBuilder().concurrencyLevel(concurrencyLevel).maximumSize(capacity).build();
+        metadataCache = CacheBuilder.newBuilder().concurrencyLevel(concurrencyLevel).maximumSize(capacity).build();
+        this.capacity = capacity;
+        this.concurrencyLevel = concurrencyLevel;
+    }
+
+    /**
+     * @return - capacity of this cache in terms of max number of entries
+     */
+    public int getCapacity() {
+        return capacity;
+    }
+
+    /**
+     * @return - concurrencyLevel of this cache,in terms of number of partitions that distinct threads can operate on
+     *         without waiting for other threads
+     */
+    public int getConcurrencyLevel() {
+        return concurrencyLevel;
+    }
+
+
+    @Override
+    public StatementPatternMetadata readStatementPatternMetadata(SnapshotBase tx, String nodeId) {
+        Optional<NodeType> type = NodeType.fromNodeId(nodeId);
+
+        try {
+            checkArgument(type.isPresent() && type.get() == NodeType.STATEMENT_PATTERN);
+            LOG.debug("Retrieving Metadata from Cache: {}", nodeId);
+            return (StatementPatternMetadata) commonNodeMetadataCache.get(nodeId, new Callable<CommonNodeMetadata>() {
```

Comment:
Could shorten these with a lambda through out this class.

```java
return (StatementPatternMetadata) commonNodeMetadataCache.get(nodeId, () -> {
    LOG.debug("Seeking Metadata from Fluo Table: {}", nodeId);
    return dao.readStatementPatternMetadata(tx, nodeId);
});
```

Reply:
Ooh.  That is elegant.  I'll keep it in mind going forward.